Implement item variants [WIP]

mugling opened the pull request:
Starts implementation of #15781

BevapDin commented on src/itype.h:
```diff
@@ -431,32 +431,61 @@ struct islot_artifact {
+
+    private:
+        template <typename T>
+        static std::unique_ptr<T> clone_slot( const std::unique_ptr<T>& src ) {
```
Instead of explicitly copying the `unique_ptr`s, you could just make a copyable `unique_ptr` and let the compiler generate the copy constructor of the whole class. For example:
```C++
template<typename T>
class copyable_unique_ptr<T> : public std::unique_ptr<T> {
    using unique_ptr<T>::unique_ptr;
    copyable_unique_ptr(const copyable_unique_ptr<T> &other)
    : unique_ptr(other ? new T(*other) : nullptr) { }
};
```

mugling replied:
That's a nice solution

mugling commented:
The major usages of `copy-from` are likely to be tools with on/off states and ammo.

24cd99d demonstrates the `copy-from` property for `9mm` ammo which includes support for a hierarchy of definitions (eg. `9mm` -> `9mmfmj` -> `reloaded_9mmfmj`)

In order to support `copy-from` the way defaults need to be applied after all possible inheritance has occurred. In df35e2d this is now changed to occur in `item_factory::finalize`.

Marked WIP as needs cleaning up and will also need to be rebased to work with #15880.

At present fields can either be replaced (or extended in the case of `flags`). Any suggestions as to the appropriate syntax for deleting fields?

BevapDin commented:
> At present fields can either be replaced (or extended in the case of  flags ). Any suggestions as to the appropriate syntax for deleting fields?

Would be nice if one could use the same syntax as for changing monster type properties: "add:\<member\>" to add to a container, "remove:\<member\>" to remove from a container and "<member>" to completely override the container.

You may even use `auto_flags_reader`, it works fine with `std::set<std::string>`.
```C++
// was_loaded should be true, copy-from was used.
optional( jsonobject, was_loaded, "member-name", member, auto_flags_reader<> {} );
```

----

> In order to support  copy-from  defaults need to be applied after all possible inheritance has occurred. In  df35e2d  this is now changed to occur in  item_factory::finalize .

I see why you need this. It has a few disadvantages:

It adds magic values (-1 meaning "no value has been set").

It separates the validity checks (weight etc.) from the loading, which means instead of "Error in foo.json:44: volume must be positive: \<relevant part of the JSON file\>", one gets "Error: \<item-id\>: unspecified or invalid volume".

However, I don't see any better solution.

----

> "cannot copy from definition of differing type"

This should definitively be possible, at least for things based on `itype` directly, not necessarily for `it_comest` / `it_tool`. Ideally, `it_comest` and `it_tool` get converted to store their additional data in an item slot instead (who says comestibles can't be tools).

Anyway, one should be able to take a briefcase (armor) and copy from it to make a briefcase SMG (gun). Or take a book and add container properties to it (hidden flask in a book). Or create an ammo type from a generic item (melee weapon).

mugling commented:
> "cannot copy from definition of differing type"
> not necessarily for it_comest / it_tool

That was indeed why I added this assertion. I see now reason we can't relax it once all of the slots support `copy-from`

> Would be nice if one could use the same syntax as for changing monster type properties: "add:<member>" to add to a container, "remove:<member>" to remove from a container and "" to completely override the container.

Can you provide an example where this idiom would be useful? I'm a little way of adding a defacto DSL to the JSON without being sure that the extra complexity is worth it?

The main use cases I have in mind are the tools with state, ammo and guns. The alternative I had in mind was implementing `copy-from` and `relative_to` with differing semantics for literals and fixed behavior for sets (`flags` always additive, `use_action` always replacement). This would obviously be less expressive than the above but it does have the advantage of simplicity whilst still covering >90% of the possible use cases.

BevapDin commented:
> Can you provide an example where this idiom would be useful? I'm a little way of adding a defacto DSL to the JSON without being sure that the extra complexity is worth it?

The "add:" and "remove:" prefixes are only a convention. "add:" and "remove:" are not parsed, the existing readers check whether there is a member "foo" and load from its content. Or (if there is no member "foo"), they check for members "remove:foo" and "add:foo" and use them instead. If you write "some-other-prefix:flags" in JSON, it will be ignored.

```JSON
{
    "id": "atomic_lamp_off",
    "copy-from": "atomic_lamp",
    "remove:flags": [ "LIGHT_100" ],
    "add:flags": ["DURABLE_MELEE"],
    "use_action": [ "turn-on-action via transform" ],
    "more tool data as needed"
}
```
"remove:flags" removes the light flag but keeps things like "LEAK_DAM" or "RADIOACTIVE" and "add:flags" adds "DURABLE_MELEE" because the cover makes the item sturdy (it protects the fragile parts).

If someone changes the flags of the base item (e.g. adds "BELT_CLIP"), those changes will be applied into the derived item. This keeps redundancy to a minimum. If needed, one can overriding the flags (using the member name without prefix) to establish an independent set of flags.

"use_action" replaces the "cover the lamp" action with a "open the cover" action (removing *all* actions of the base item).

One can make an item that can be opened and has more iuse functions when opened, but less flags.

And this system can be reused by mods to change existing items.

> The main use cases I have in mind are the tools with state, ammo and guns. The alternative I had in mind was implementing  copy-from  and  relative_to  with differing semantics for literals and fixed behavior for sets ( flags  always additive,  use_action  always replacement). This would obviously be less expressive than the above but it does have the advantage of simplicity whilst still covering >90% of the possible use cases.

That would affect all properties at once, which seems inflexible (see example). Also, this (different behavior for  "copy-from" and "relative_to") could be considered a DSO just as well.

Also: can it be used to add *and* remove flags, while keeping flags that are not explicitly remove (like in the example)?

mugling commented:
> And this system can be reused by mods to change existing items.

That's the major advantage of your approach. Changes in mainline would however then propagate to mods but whether that's a disadvantage is debatable and such behavior shouldn't catch anyone by surprise.

> Also: can it be used to add and remove flags, while keeping flags that are not explicitly remove

No, that's an implicit limitation. I'm not sure there's as many cases where that's required compared to the other widespread duplication that can be avoided (in particular the ammo).

The two approaches aren't necessarily mutually exclusive. I might re-implement `relative-to` as `relative:*` as that has a lot of potential usages.

07993ca demonstrates `relative-to`:
```
"id": "reloaded_9mmP2",
"relative-to": "9mmP2",
"type": "AMMO",
"name": "reloaded 9x19mm +P+",
"description": "A step beyond the high-pressure 9mm ... has been hand loaded",
"damage": -1,
"dispersion": 20,
"effects": [ "RECYCLED" ]
```

Balanced against both `9mm` and `9mmP2` without the use of magic values with everything in the same JSON file. Implementing hand loaded ammo is however now much less painful. Mod authors can now implement new items using raw values or derive from existing via `copy-from` or `relative-to`.

I quite like @Coolthulhu's from #15876
> Then you could implement it as variants - base item would be an abstract "pistol" or "rifle"

It's tempting to define a series of such items under `classes` (avoiding load-order shenanigans)

mugling commented:
On further consideration @BevapDin's proposal is cleaner. For any literals there's no difference between the approaches but for anything more complex an appropriate default behavior is ambiguous. I've gone with an (initial) implementation of the `add:*`, `remove:*` and `relative:*` prefix operators in 2d4f7ca

mugling commented:
New syntax following 07993ca is:
```
"id": "reloaded_9mmP2",
"copy-from": "9mmP2",
"type": "AMMO",
"name": "reloaded 9x19mm +P+",
"description": "A step beyond the high-pressure 9mm ... has been hand loaded",
"relative:damage": -1,
"relative:dispersion": 20,
"add:effects": [ "RECYCLED" ]
```

kevingranade commented on src/init.cpp:
```diff
@@ -137,31 +137,31 @@ void DynamicDataLoader::initialize()
 
     type_function_map["trap"] = new StaticFunctionAccessor(&trap::load);
     type_function_map["AMMO"] = new ClassFunctionAccessor<Item_factory>(item_controller,
-            &Item_factory::load_ammo);
```
This subverts the intent of how this is dispatched very badly :/

kevingranade commented on src/item_factory.cpp:
```diff
@@ -721,59 +767,103 @@ void Item_factory::load( islot_software &slot, JsonObject &jo )
-    if( !jo.read( "loudness", slot.loudness ) ) {
-        slot.loudness = std::max( std::max( { slot.damage, slot.pierce, slot.range } ) * 3, slot.recoil / 3 );
+    itype *src = nullptr;
+    if( jo.has_string( "copy-from" ) ) {
```
How about wrapping this in a method and having each of the load_* methods call it instead of making dispatch all crazy?

mugling replied:
The main obstacle and reason for this decision was `it_tool` and `it_comest` which are causing other problems as identified by @BevapDin. It would seem reasonable to drop both of these derived classes and then go with a method at the start of each `load_*` call

kevingranade replied:
I agree modelling all of these as capabilities with slots is better than subclassing.

mugling replied:
See #15901

kevingranade commented:
This adds a very strong and very random ordering dependency on json loading, which I've been trying to avoid for as long as we've had json loading.  Can you defer the "copy-from" functionality to the finalize step?

kevingranade commented on data/json/items/ammo/9mm.json:
```diff
@@ -23,164 +23,84 @@
-    "recoil": 195,
-    "effects": [ "COOKOFF" ]
-
+    "relative:damage": -2,
```
Please don't do manual tokenizing like this, just use json.
```"relative" : { "damage" : -2, "pierce" : 2 }```

mugling replied:
With the corresponding syntax?
```
"add": { "flags" : "RECYCLED" }
"delete": { "flags": "FOO" }
```

kevingranade replied:
Yep, just so.

mugling commented:
> Can you defer the "copy-from" functionality to the finalize step?

Not easily. The difficulty being each `itype` needs to remember for every member or slot whether a value has been set versus left as the default. This would require a lot of magic values (`INT_MIN`, `-1` etc) and most/all of the logic to be pushed into `finalize()`.

In the long-term I'd like to rework JSON parsing to support this but for now the aim is to implement the core functionality to support ammo and stateful tools. Such a future extension should probably include warnings for unused fields (typically typos) and forbid specifying the default value.

kevingranade commented:
There's a slightly messier (but at the same time cleaner, go figure) way to do this, which is to defer parsing of a json blob that inherits from another json entity until finalization.  You can simply stash a string representing an entities' json definition on a separate list of objects, then when finalize() is called iterate over that list evaluating it until all of them resolve.
psudocode:
```
on load handler:
  if( jo.has_string( "copy_from" ) &&
      m_templates.find( jo.get_string( "copy-from" ) ) == m_templates.end() ) {
    deferred_entities.push_back( jo.contents() );
    return;
  }
  // Normal JSON load...  

in finalize:
  // This will continue looping over the deferred entities until they run out,
  // load_all_from_json() will keep re-inserting them if they can't find their dependencies.
  // It needs to additionally report when that happens so the loop can detect when it has stalled.
  while( !deferred_entities.empty() ) {
    JsonIn jsin( deferred_entities.[0] );
    load_all_from_json(jsin);
  }
```

mugling commented:
> ou can simply stash a string representing an entities' json definition on a separate list of objects, then when finalize() is called iterate over that list evaluating it until all of them resolve

Needs a slight variation to break circular dependencies but I don't otherwise see a problem with this solution?

kevingranade commented:
Circular dependencies are simply unresolveable in this situation, so all
you can do is detect them and fail.

mugling commented:
> Circular dependencies are simply unresolveable in this situation, so all you can do is detect them and fail.

Indeed

mugling commented:
Continued in #15917
